qlib/contrib/model/pytorch_hgats_ts.py

```diff
@@ -0,0 +1,547 @@
+            )  # [#stocks, #industries]
+
+            with torch.no_grad():
+                pred = (
+                    self.HGAT_model(
+                        feature.float(), torch.Tensor(GH.values).to(self.device)
+                    )
+                    .detach()
+                    .cpu()
+                    .numpy()
+                )
+
+            preds.append(pred)
+
+        return pd.Series(
+            np.concatenate(preds),
+            index=dl_test.get_index()[: len(np.concatenate(preds))],
+        )
+
+
+class HGATModel(nn.Module):
+    def __init__(
+        self,
+        d_feat=6,
+        hidden_size=64,
+        num_layers=2,
+        dropout=0.0,
+        num_ind=29 + 1,  # number of industries plus an 'unknown' one
+        base_model="GRU",
+    ):
+        super().__init__()
+
+        if base_model == "GRU":
+            self.rnn = nn.GRU(
+                input_size=d_feat,
+                hidden_size=hidden_size,
+                num_layers=num_layers,
+                batch_first=True,
+                dropout=dropout,
+            )
+        elif base_model == "LSTM":
+            self.rnn = nn.LSTM(
+                input_size=d_feat,
+                hidden_size=hidden_size,
+                num_layers=num_layers,
+                batch_first=True,
+                dropout=dropout,
+            )
+        else:
+            raise ValueError("unknown base model name `%s`" % base_model)
+
+        self.hidden_size = hidden_size
+        self.num_ind = num_ind
+        self.d_feat = d_feat
+        self.transformation = nn.Linear(self.hidden_size, self.hidden_size)
+        self.a = nn.Parameter(torch.randn(self.hidden_size * 2, 1))
+        self.a.requires_grad = True
+        self.fc = nn.Linear(self.hidden_size, self.hidden_size)
+        self.fc_out = nn.Linear(hidden_size, 1)
+        self.leaky_relu = nn.LeakyReLU()
+        self.softmax = nn.Softmax(dim=1)
+
+    def cal_attention(self, x, y):
+        x = self.transformation(x)  # x <- hidden
+        y = self.transformation(y)  # y <- hidden_agg
+
+        sample_num = x.shape[0]  # num of stocks
+        edge_num = y.shape[0]  # num of industries
+        dim = x.shape[1]  # num of temporal features
+        e_x = x.expand(edge_num, sample_num, dim)
+        e_y = y.expand(sample_num, edge_num, dim)
+        e_y = torch.transpose(e_y, 0, 1)
+        attention_in = torch.cat((e_x, e_y), 2).view(-1, dim * 2)  # P i || P j
+        self.a_t = torch.t(self.a)
+        attention_out = (
+            self.a_t.mm(torch.t(attention_in)).view(edge_num, sample_num).t()
+        )
+        attention_out = self.leaky_relu(attention_out)
+        att_weight = self.softmax(attention_out)
+
+        return att_weight
+
+    def forward(self, x, GH):
+
+        out, _ = self.rnn(x)
+        hidden = out[:, -1, :]  # [#stocks, #features]
+
+        hidden_agg = torch.t(GH).mm(hidden)  # [#industries, #features]
```

Comment:
The aggregation is a little weird. 
The scale is not aligned and the meaning of  `hidden_agg`  may be greatly different from hidden (But they will be transformed by the same `self.transformation` later)

Reply:
The meaning of `hidden_agg` and `hidden` are different by design. `hidden_agg` corresponds to `xj` in the paper and `hidden` corresponds to `xi`. They go through the same linear transformation `P`.

<img width="471" alt="Screen Shot 2021-12-10 at 11 30 56" src="https://user-images.githubusercontent.com/50850474/145512668-258bbdde-530d-40f9-bf32-6221f509fb61.png">